src/java.base/share/classes/java/lang/invoke/X-VarHandleMemoryAccess.java.template

```diff
@@ -66,305 +77,311 @@ final class MemoryAccessVarHandle$Type$Helper {
 #end[floatingPoint]
 
     @ForceInline
-    static MemoryAddressProxy checkAddress(Object obb, long offset, long length, boolean ro) {
-        MemoryAddressProxy oo = (MemoryAddressProxy)Objects.requireNonNull(obb);
+    static MemorySegmentProxy checkAddress(Object obb, long offset, long length, boolean ro) {
+        MemorySegmentProxy oo = (MemorySegmentProxy)Objects.requireNonNull(obb);
         oo.checkAccess(offset, length, ro);
         return oo;
     }
-    
+
     @ForceInline
-    static long offset(MemoryAddressProxy bb, long offset, long alignmentMask) {
-        long address = offsetNoVMAlignCheck(bb, offset, alignmentMask);
+    static long offset(boolean skipOffsetCheck, MemorySegmentProxy bb, long offset, long alignmentMask) {
+        long address = offsetNoVMAlignCheck(skipOffsetCheck, bb, offset, alignmentMask);
         if ((address & VM_ALIGN) != 0) {
             throw MemoryAccessVarHandleBase.newIllegalStateExceptionForMisalignedAccess(address);
         }
         return address;
     }
 
     @ForceInline
-    static long offsetNoVMAlignCheck(MemoryAddressProxy bb, long offset, long alignmentMask) {
+    static long offsetNoVMAlignCheck(boolean skipOffsetCheck, MemorySegmentProxy bb, long offset, long alignmentMask) {
         long base = bb.unsafeGetOffset();
         long address = base + offset;
-        //note: the offset portion has already been aligned-checked, by construction
-        if ((base & alignmentMask) != 0) {
-            throw MemoryAccessVarHandleBase.newIllegalStateExceptionForMisalignedAccess(address);
+        if (skipOffsetCheck) {
+            //note: the offset portion has already been aligned-checked, by construction
+            if ((base & alignmentMask) != 0) {
+                throw MemoryAccessVarHandleBase.newIllegalStateExceptionForMisalignedAccess(address);
+            }
+        } else {
+            if ((address & alignmentMask) != 0) {
+                throw MemoryAccessVarHandleBase.newIllegalStateExceptionForMisalignedAccess(address);
+            }
```

Comment:
The final address is the sum of two components - a base B and an offset O. In a hot loop, B is typically constant (e.g. the base address of a segment) while O keeps changing. So, if the alignment check is only performed on B, the VM can prove that this check can be hoisted outside the loop. Of course that leaves the correctness problem - e.g. what if B is aligned but B + O is not? Luckily, by construction, this is not possible - skipOffsetCheck is only set when we construct a memory access var handle from a layout - and in that case the layout API make sure that the O part of the final address is aligned accordingly.

So, while it "looks" as if there's no performance gain, in reality the gain is quite big, because that move allows the alignment check to be moved outside the loop.

Reply:
Thanks for the explanation